nixos/modules/services/networking/smartdns.nix

```diff
@@ -0,0 +1,62 @@
+{ lib, pkgs, config, ... }:
+
+with lib;
+
+let
+  cfg = config.services.smartdns;
+  mkServerConfig = method:
+    mkOption {
+      type = types.listOf types.str;
+      default = [ ];
+      description = ''
+        List of servers to query with ${toUpper method} method. You can apply different options at here too. See the <link xlink:href="https://github.com/pymumu/smartdns/blob/master/ReadMe_en.md#configuration-parameter">SmartDNS README</link> for details.
+      '';
+    };
+
+  confFile = pkgs.writeText "smartdns.conf" ''
+    bind :${toString cfg.bindPort}
+    cache-size ${toString cfg.cacheSize}
+    ${flip concatMapStrings [ "udp" "tcp" "tls" "https" ] (method:
+      flip concatMapStrings cfg.servers.${method} (server: ''
+        server-${method} ${server}
+      ''))}
+    ${cfg.extraConfig}
+  '';
+in {
+  options.services.smartdns = {
+    enable = mkEnableOption "SmartDNS DNS server";
+
+    servers = genAttrs [ "udp" "tcp" "tls" "https" ]
+      (method: mkServerConfig "${method}");
+
+    cacheSize = mkOption {
+      type = types.int;
+      default = 512;
+      description = "Domain name result cache number.";
+    };
+
+    bindPort = mkOption {
+      type = types.int;
+      default = 53;
+      description = "DNS listening port number.";
+    };
+
+    extraConfig = mkOption {
+      type = types.lines;
```

Comment:
Please use the approach described in NixOS/rfcs#42 instead, since it provides multiple benefits for the future. That is:
- Replace this option with a `settings` option that has a type of
  ```nix
  attrsOf (oneOf [ str int bool ])
  ```
  instead, representing the config as a Nix value, not a string.
- In the `config` section of the module, specify the default values by setting e.g.
  ```nix
  {
    services.smartdns.settings.bind = mkDefault ":${toString cfg.bindPort}";
  }
  ```
- Generate the config file with a generator function from the `settings` option, like this:
  ```nix
  let
    confFile = pkgs.writeText "smartdns.conf" (
      with generators; toKeyValue {
        mkKeyValue = mkKeyValueDefault {
          mkValueString = v:
            if isBool v then if v then "yes" else "no"
            else mkValueStringDefault {} v;
        } " ";
      } cfg.settings);
  in /* ... */
  ```

Reply:
Also an example for this option would then be great, showing a bunch of different and useful settings, it would look like
```nix
{
  example = literalExample ''
    {
      some-key = "some-value";
    }
  '';
}
```

Reply:
Not sure how would this work with multiple bound ports (multiple `bind` directives). The config might look something like this:

```
Bind :7053 -group office
Bind :8053 -group home
Server 192.168.1.1 -group home -exclude-default-group
Server 10.0.0.1 -group office -exclude-default-group
Server 8.8.8.8
```

Reply:
Oh, that's lucky, because I just made https://github.com/NixOS/nixpkgs/pull/82208 to support exactly that (by setting lists of values)!

So you can just pass `listsAsDuplicateKeys = true` as an argument to `toKeyValue` to make this work, and change the type to `let atom = oneOf [ str int bool ]; in attrsOf (either atom (listOf atom))`

Reply:
Thanks for comment, I will change soon!

Reply:
@Infinisil Well, I agree that using universal `settings` would be a good idea. But for the [configuration parameters](https://github.com/pymumu/smartdns/blob/master/ReadMe_en.md#configuration-parameter) here, I have no idea about how would this be represented in a Nix module. The format here is very CLI-args-like but not key-value pair like.

For example, `bind :53 -no-rule -group example` could be possible, which means `bind = 53` is not enough to represent.

Reply:
Just `bind = ":53 -no-rule -group example"` should be fine

Reply:
> Just `bind = ":53 -no-rule -group example"` should be fine

OK. Although it looks a little ugly.

Reply:
Correction: In order for lists to work the type needs to be
```nix
let atom = oneOf [ str int bool ];
in attrsOf (coercedTo atom toList (listOf atom))
```